---
title: "Inferring Breast Cancer Risk with Genetic Data"
author: "Julian McClellan"
date: "8/25/2020"
output: 
  revealjs::revealjs_presentation:
    theme: night
    transition: convex
---

```{r get_cohort_info, include=F}
knitr::opts_chunk$set(include=F, echo=F)
# Get IDs for patients
elai.inp.geno <- c("../data/CancerChr22.geno.txt", "../data/CancerChr22.geno.txt")

makeTable <- function(df, font_size){
  require(dplyr)
  df %>%
    rmarkdown::paged_table() %>%
    knitr::kable("html",
                 digits = 5) %>%
    kable_styling(bootstrap_options = "striped",
                  full_width = F, font_size = font_size)
}

extractThirdLine <- function(fp){
  require(stringr)
  
  con <- file(fp, "r")
  lines <- c()
  gotThird <- F
  
  count <- 0
  while(count != 3) {
    line <- readLines(con, 1)
    count <- count + 1
  }
  
  close(con)
  ids <- str_split(line, ", ", simplify=T)[-1] # Leave out title "IND"
  return(ids)
}

ids.chr21 <- extractThirdLine(elai.inp.geno[1])
# ids.chr22 <- extractThirdLine(elai.inp.geno[2])

# IDs should be in the same order
# all(ids.chr21 == ids.chr22) 

## Get cases

chr21.fam <- read.table("../data/CIDR_Olopade_Plus_hg19_3686subject_chr21.fam")
# chr22.fam <- read.table("../data/CIDR_Olopade_Plus_hg19_3686subject_chr22.fam")

# all(chr21.fam$V6 == chr21.fam$V6) # Cases should be same
# all(chr21.fam$V2 == chr21.fam$V2) # ==> IDs should be same
# all(chr21.fam$V2 == ids.chr21)
# all(chr21.fam$V2 == ids.chr22)

library(dplyr)
cohort <- tibble(ind = ids.chr21, case = chr21.fam$V6) %>%
  mutate(status = case_when(case == 2 ~ "cancer",
                                   case == 1 ~ "control",
                                   TRUE ~ "unknown"))
snpinfo.chr21 <- read.table("../output/Adm_Cancer_Chr21.snpinfo.txt", skip=2)
snpinfo.chr22 <- read.table("../output/Adm_Cancer_Chr22.snpinfo.txt", skip=2)

rsids.chr21 <- snpinfo.chr21[,1]
rsids.chr22 <- snpinfo.chr22[,1]
```

```{r get_training_data, eval=F, include=F}
library(dplyr)
rsids.chr21 <- read.table("../output/Adm_Cancer_Chr21.snpinfo.txt", skip=2)[,1]
rsids.chr22 <- read.table("../output/Adm_Cancer_Chr22.snpinfo.txt", skip=2)[,1]

yy.chr21 <- scan("../output/Adm_Cancer_Chr21.ps21.txt")
S.chr21 = 2 # (source populations)
M.chr21 = 20352 # (# SNPs)
N.chr21 = 3686 # (individuals in sample)
dim(yy.chr21) <- c(S.chr21, M.chr21, N.chr21)

# Equivalent way of getting "sy" in elai example, 2 for YRI (African ancestry)
yy.chr21 <- as_tibble(t(yy.chr21[2,,]))
names(yy.chr21) <- rsids.chr21

yy.chr22 <- scan("../output/Adm_Cancer_Chr22.ps21.txt")
S.chr22 = 2 # (source populations)
M.chr22 = 21244 # (# SNPs)
N.chr22 = 3686 # (individuals in sample)
dim(yy.chr22) <- c(S.chr22, M.chr22, N.chr22)

yy.chr22 <- as_tibble(t(yy.chr22[2,,]))
names(yy.chr22) <- rsids.chr22

# Stick allele dosages for chr21 and chr22 SNPs together for 3686 cohort individuals
yy.chr2122 <- bind_cols(cohort, yy.chr21, yy.chr22)

yy.chr21 <- yy.chr21 %>%
  bind_cols(cohort, .)

yy.chr22 <- yy.chr22 %>%
  bind_cols(cohort, .)


saveRDS(yy.chr2122, "../data/for_R/chr2122.rds")
saveRDS(yy.chr21, "../data/for_R/chr21.rds")
saveRDS(yy.chr22, "../data/for_R/chr22.rds")

rm(list = ls())
```

```{r display_cohort}
library(kableExtra)

cohort_breakdown <- cohort %>%
  group_by(status) %>%
  summarize(n = n()) %>%
  makeTable(font_size=26)

snps.chr21 <- length(rsids.chr21)
snps.chr22 <- length(rsids.chr22)
snp_breakdown <- tibble(chromosome = c(21, 22),
                        SNPs = c(snps.chr21, 
                                 snps.chr22)) %>% 
  makeTable(font_size=26)
```

```{r get_panel_info}
ids.ceu <- extractThirdLine("../data/1000CEU_Chr21.geno.txt")
ids.yri <- extractThirdLine("../data/1000YRI_Chr21.geno.txt")

# ids.ceu22 <- extractThirdLine("../data/1000CEU_Chr22.geno.txt")
# ids.yri22 <- extractThirdLine("../data/1000YRI_Chr22.geno.txt")

panel_breakdown <- bind_rows(
  tibble(id = ids.ceu, ancestry = "Northern/Western Europe"),
  tibble(id = ids.yri, ancestry = "West Africa (Nigeria)")
) %>%
  group_by(ancestry) %>%
  summarize(n = n()) %>%
  makeTable(font_size=26)
```

# Overview

<p style="text-align:left;">Using **E**fficient **L**ocal **A**ncestry **I**nference (**ELAI**) we will estimate SNP allele dosages for a sample of African Americans using source populations from West Africa and Northern/Western Europe.</p>

<p style="text-align:left;">Combining the ELAI results with the cancer statuses of the sample individuals, we will fit a model to infer breast cancer risk.</p>

# Data

- `r nrow(cohort)` individuals in the sample cohort 
- `r length(ids.ceu) + length(ids.yri)` individuals in the reference panel

## Cohort

`r cohort_breakdown`

## Panel

`r panel_breakdown`

# Transformations

<p style="text-align:left;">Both the cohort files and panel files end up as bimbam format files for entry use in ELAI.</p>
 
## Sample preprocessing

![Sample preprocessing](img/preprocess_sample.png)

## Reference Preprocessing

![Reference preprocessing](img/preprocess_reference.png)

## Bimbam Processing

 - <p style="text-align:left;">For the `*geno.txt` files for the cohort and panel, the number of SNPs must be put on the 2nd row.</p>
 
![](img/bimbam_edit1.png)

## Bimbam Processing

 - <p style="text-align:left;">For the sample position files the chromsome ID is added.</p>
 
![](img/bimbam_edit2.png)
 
## Bimbam Processing

 - <p style="text-align:left;">For the reference `*geno.txt` files, we add an `"="` after the number of individuals because these genotypes are *phased*, that is, we can identify which chromosome the genotypes' alleles are from.</p>

![](img/bimbam_edit3.png)

# Ancestry Inference

<p style="text-align:left;">ELAI uses a two-layer Hidden Markov model to infer local ancestry of admixed individuals.</p>

<p style="text-align:left;">Given our sample of African Americans . . .</p>

 - African Americans have 2 source populations:
    - West Africans
    - Northern/Western Europeans
 - These 2 source populations began to mix ~10 generations ago.
 
<p style="text-align:left;">These facts inform our final ELAI command.</p>
 
## ELAI Command

![ELAI Command](img/infer_ancestry.png)

## ELAI Output | Dosage

- A file containing a table with dimensions $(S, M, N)$, where:
  - $S$ corresponds to the number of source populations (`2`)
  - $M$ to the number of SNPs present in all position and genotype files.
 `r snp_breakdown`
  - $N$ to the number of admixed individuals in the sample (`3686`).
- For a given $s$, $m$, and $n$ in the file we will see the expected number of alleles from the $s$th source population for $m$th marker for  $n$th individual. (allele dosage $\in[0,2]$)
  
## ELAI Output | SNP Info

- A file showing information on the SNPs.
  - `rsID`
  - `minor allele`
  - `major allele`
  - `minor allele frequency`
  - `major allele frequency`
  - `position`
  
# Cancer Risk Inference

<p style="text-align:left;">By combining elements of the ELAI input and output, we can create a file suitable for cancer risk inference, a $N \times M$ file where individual cancer statuses and rsIDs of SNPs can be identified.</p>

<p style="text-align:left;">So for a given chromsome(s) with $M$ SNPs we model an individual $i$'s cancer status as:</p>

<br>
<br>

$$cancer\_status_{i} \sim allele\_dosage_{i1} + \ \cdot\cdot\cdot \ + allele\_dosage_{iM}$$

## Logistic LASSO Regression

- LASSO has the nice property of zeroing out coefficients.
  - Can allow a simpler model that focuses on the most relevant SNP predictors.
- Model selection chooses the model with highest ROC using: 
  - Scaled data (all dosages have `SD` of `1`).
  - `90%` of the data is utilized for training, with `10%` reserved to assess model performance.
  - 5 fold cross validation repeated 50 times across 100 possible lambda values (equidistant on log scale from `.001` to `100`).

<br>
Models

1. `r snps.chr21` SNPs from Chromosome 21
2. `r snps.chr22` SNPs from Chromosome 22
3. `r snps.chr21 + snps.chr22` SNPs from Chromosomes 21 + 22 (separate ELAI runs)

## Performance | Final Models | Held Out Test Data

![](cluster_output/models.perf.png){width=70%}

## Performance | ROC Bootstrap 95% CIs

```{r model_coefs, include=T, message=F, echo=F}
library(readr)

getCoefs <- function(sparse_coefs){
  require(dplyr)
  require(kableExtra)
  snp_names <- row.names(sparse_coefs)
  
  df_coefs <- as.data.frame(as.matrix(sparse_coefs)) %>%
    mutate(snp = snp_names) %>%
    rename(coef = `1`) %>%
    arrange(coef) %>%
    filter(coef != 0)
  
  return(df_coefs)
}

model.21.coefs <- read_rds("cluster_output/model.21.coefs.rds") %>% 
  getCoefs() %>%
  left_join(snpinfo.chr21[,c(1,6)], by = c("snp" = "V1")) %>%
  rename(position = V6) %>%
  arrange(coef)
model.22.coefs <- read_rds("cluster_output/model.22.coefs.rds") %>%
  getCoefs() %>%
  left_join(snpinfo.chr22[,c(1,6)], by = c("snp" = "V1")) %>%
  rename(position = V6) %>%
  arrange(coef)
model.2122.coefs <- read_rds("cluster_output/model.2122.coefs.rds") %>%
  getCoefs() %>%
  left_join(bind_rows(snpinfo.chr22[,c(1, 5, 6)],
                      snpinfo.chr21[,c(1, 5, 6)])
                      , by = c("snp" = "V1")) %>%
  rename(position = V6, chromosome = V5) %>%
  arrange(chromosome, coef)


models.overlap.coefs <- left_join(
  left_join(
  model.2122.coefs,
  model.21.coefs %>% mutate("In 21 Model" = "Yes")),
  model.22.coefs %>% mutate("In 22 Model" = "Yes")
) %>% 
  arrange(chromosome, position, coef) %>%
  mutate(`In 21 Model` = case_when(is.na(`In 21 Model`) ~ "No",
                             TRUE ~ "Yes" 
                             ),
         `In 22 Model` = case_when(is.na(`In 22 Model`) ~ "No",
                             TRUE ~ "Yes"
                             )
         )

table.model.22.coefs <- model.22.coefs %>% makeTable(13)
table.model.21.coefs <- model.21.coefs %>% makeTable(14)
table.model.2122.coefs <- model.2122.coefs %>% makeTable(12)
table.models.overlap.coefs <- models.overlap.coefs %>%
  makeTable(12)

(roc.plot <- read_rds("cluster_output/roc_plot.rds"))

models.tune <- read_rds("cluster_output/models.bestTune.rds")

tune_table <- data.frame(model = c("21", "22", "21/22"),
                     lambda = c(models.tune$model.21$lambda,
                                models.tune$model.22$lambda,
                                models.tune$model.2122$lambda),
                     SNPs = c(snps.chr21,
                              snps.chr22,
                              snps.chr21 + snps.chr22
                              )
) %>%
  makeTable(26)
```

## Lambda Parameters

`r tune_table`

## Coefficients | Chromsome 21

`r table.model.21.coefs`

## Coefficients | Chromsome 21 {data-transition="none"}

```{r plot_chr21_coefs, include=T}
plot.xlims <- c(min(snpinfo.chr21$V6, snpinfo.chr22$V6), max(snpinfo.chr21$V6, snpinfo.chr22$V6))

library(ggplot2)
(model.21.plot <- model.21.coefs %>%
  filter(snp != "(Intercept)") %>%
  mutate(y = 1) %>%
  ggplot(aes(x=position, y=coef, alpha=.5)) + 
    geom_point(color="#F8766D") + 
    xlim(plot.xlims[1], plot.xlims[2]) + 
    guides(alpha=F) +
    # labs(title = "Model Coefficients", subtitle="Chromosome 21 Only") +
    geom_hline(yintercept = 0, linetype=3) +
    theme_minimal())
```

## Coefficients | Chromsome 22

`r table.model.22.coefs`

## Coefficients | Chromsome 22 {data-transition="none"}

```{r plot_chr22_coefs, include=T}
library(ggplot2)
(model.22.plot <- model.22.coefs %>%
  filter(snp != "(Intercept)") %>%
  ggplot(aes(x=position, y=coef, alpha=.5)) + 
    geom_point(color="#00BA38") + 
    xlim(plot.xlims[1], plot.xlims[2]) + 
    guides(alpha=F) +
    # labs(title = "Model Coefficients", subtitle="Chromsome 22 Only") +
    geom_hline(yintercept = 0, linetype=3) +
    theme_minimal())
```

## Coefficients | Chromsome 21/22

`r table.model.2122.coefs`

## Coefficients | Chromsome 21/22 {data-transition="none"}

```{r plot_chr2122_coefs, include=T}
library(ggplot2)
model.2122.coefs$chromosome <- factor(model.2122.coefs$chromosome)
(model.2122.plot <- model.2122.coefs %>%
  filter(snp != "(Intercept)") %>%
  ggplot(aes(x=position, y=coef, alpha=.5)) + 
    geom_point(color="#619CFF") + 
    xlim(plot.xlims[1], plot.xlims[2]) + 
    guides(alpha=F, color=F) +
    # labs(title = "Model Coefficients", subtitle="Chromsomes 21 and 22") +
    geom_hline(yintercept = 0, linetype=3) +
    facet_grid(chromosome ~.) +
    theme_minimal())
```

## Coefficients | Overlap?

`r table.models.overlap.coefs`

## Coefficients | Overlap? {data-transition="none"}

```{r plot_chr2122_overlap_coefs, include=T}
name.chr21 <- "Chr21     "
name.chr22 <- "Chr22     "

model.21.coefs.comp <- model.21.coefs %>%
  mutate(model=name.chr21, chromosome=as.factor(21))

model.22.coefs.comp <- model.22.coefs %>%
  mutate(model=name.chr22, chromosome=as.factor(22))

model.2122.coefs.comp <- model.2122.coefs %>%
  mutate(model="Chr21/22")

ggdata.overlap <- bind_rows(model.21.coefs.comp, model.22.coefs.comp, model.2122.coefs.comp)
ggdata.overlap$model <- factor(ggdata.overlap$model, levels=c(name.chr21, name.chr22, "Chr21/22"))

ggdata.2122.overlap <- ggdata.overlap %>%
  filter(model %in% c(name.chr21, name.chr22))


library(ggplot2)
(models.overlap.plot <- ggdata.overlap %>%
  filter(snp != "(Intercept)") %>%
  ggplot(aes(x=position, y=coef, alpha=.5, color=model)) + 
    geom_point() + 
    xlim(min(snpinfo.chr21$V6, snpinfo.chr22$V6), max(snpinfo.chr21$V6, snpinfo.chr22$V6)) + 
    guides(alpha=F) +
    # labs(title = "Models Coefficient Comparison") +
    geom_hline(yintercept = 0, linetype=3) +
    facet_grid(chromosome ~.) +
    theme_minimal())
```

## Coefficients | Overlap? {data-transition="none"}

```{r models2122_overlap_plot, include=T}
(models.2122.overlap.plot <- ggdata.2122.overlap %>%
  filter(snp != "(Intercept)") %>%
  ggplot(aes(x=position, y=coef, alpha=.5, color=model)) + 
    geom_point() + 
    scale_color_manual(values=c("#F8766D", "#00BA38")) +
    xlim(min(snpinfo.chr21$V6, snpinfo.chr22$V6), max(snpinfo.chr21$V6, snpinfo.chr22$V6)) + 
    guides(alpha=F) +
    # labs(title = "Models Coefficient Comparison") +
    geom_hline(yintercept = 0, linetype=3) +
    facet_grid(chromosome ~.) +
    theme_minimal())
```

# Potential Adjustments: Transformation

- Could do file format transformations without `fcgene`.
  - `plink1.9` offers bimbam output that differs slightly from `fcgene's`, but
  should still work in ELAI.
  
## Potential Adjustments: Ancestry Inference

- ELAI author recommends averaging results over different mixture generation numbers. 
  - Could also weight by uncertainty.
  - Joint model fitting: Running additional EM steps on source *and* admixed samples is also recommended by the author.
- Could increase # of EM steps above 30.
- ELAI can be run with SNPs from multiple chromosomes simultaneously. 
  - Some further adjustment of ELAI input position files would be necessary.

## Potential Adjustments: Cancer Risk Inference

- Various different models, feature selection, parameter tuning schemes, etc.
- Utilize selected SNPs from LASSO in plain logistic regression (to get SE's for coefficients)

## Potential Adjustments: Overall

- Refactoring code.
  - Combine with virtualization to make analysis reproducible and portable.
  
# End

```{r toy_model, eval=F}
library(readr)
library(dplyr)

train_model <- function(full_dat, trim, toy=T){
  require(caret)
  require(dplyr)
  # require(doParallel)
  # require(doMC)

  a <- seq(log(0.001), log(100), length.out = 25)
  b <- exp(a) # Left biased lambda search

  # Lasso lambda params
  hyper_grid <- expand.grid(alpha = 1, lambda = b)

  full_dat <- full_dat %>%
    filter(status != "unknown") %>%
    select(-c(ind, case))
  
  if (toy == T){
    full_dat <- full_dat[c(1:1000), c(1:100)]
  }

  if (trim == 1){ # Most trimmed/smallest model (smallest in memory)
    ctrl <- trainControl(method = "cv", # 10 folds default
                         # Save memory
                         trim = T,
                         returnData = F,
                         classProbs = T,
                         savePredictions = T,
                         summaryFunction = twoClassSummary)

  } else if (trim == 2) {
    ctrl <- trainControl(method = "cv",
                         returnData = T,
                         trim = T,
                         classProbs = T,
                         savePredictions = T,
                         summaryFunction = twoClassSummary)
  
  } else if (trim == 3) { # Fattest model (biggest in memory)
    ctrl <- trainControl(method = "repeatedcv",
                         repeats = 5,
                         returnData = T,
                         classProbs = T,
                         savePredictions = T,
                         summaryFunction = twoClassSummary)
  }

  options(expressions = 500000)
  # registerDoMC(2)
  # Nslots <- as.numeric(Sys.getenv("SLURM_CPUS_ON_NODE"))
  # cat(as.character(Nslots), "\n")
  # cl <- makeCluster(Nslots, outfile = "")
  #
  # registerDoParallel(cl)
  #

  set.seed(8162020)
  model <- train(status ~ .,
                 data = full_dat,
                 method = "glmnet",
                 metric = "ROC",
                 trControl = ctrl,
                 tuneGrid = hyper_grid
  )
}

train_model(read_rds("../data/for_R/chr21.rds"), 3) %>%
  write_rds("../output/model.chr21.rds")
train_model(read_rds("../data/for_R/chr22.rds"), 3) %>%
  write_rds("../output/model.chr22.rds")
train_model(read_rds("../data/for_R/chr2122.rds"), 3) %>%
  write_rds("../output/model.chr2122.rds")
```
```{r toy_present, eval=F}
# model.most.skinny.21 <- read_rds("../output/model.chr21.skinny.most.rds")
# model.skinny.21 <- read_rds("../output/model.chr21.skinny.rds")
model.21 <- read_rds("../output/model.chr21.rds")
model.22 <- read_rds("../output/model.chr22.rds")
model.2122 <- read_rds("../output/model.chr2122.rds")

perf <- MLeval::evalm(list(model.21, model.22, model.2122), title = "ROC Curve Comparison", gnames=c("Chromosome 21", "Chromosome 22", "Chromosomes 21/22"), showplots = F, plots = "r", rlinethick = 1)

# save_rds(perf, "../output/performance.rds")

# coefplot <- coefplot::multiplot(list(`Chromosome 21` = model.21$finalModel, `Chromosome 22` = model.22$finalModel, `Chromosome 21/22` = model.2122$finalModel),
#                                 pointSize = 1.75) +
#   theme_minimal()
```
